### ISSUE TITLE
Problems with inherited associated types

### ISSUE THREAD

**Issue by ilyavenner** [dtolnay/async-trait#61]
Hello there. I have problem with inherited associated types. The code below compiles successfully:
```rust
struct Foo;

trait T1 {
    type T;

    fn f1(&self, t: Self::T) -> Self::T;
}

trait T2: T1 {
    fn f2(&self, t: Self::T) -> Self::T;
}

impl T1 for Foo {
    type T = Foo;

    fn f1(&self, t: Self::T) -> Self::T {
        t
    }
}

impl T2 for Foo {
    fn f2(&self, t: Self::T) -> Self::T {
        self.f1(t)
    }
}
```
but if I add `#[async_trait]` like in this example:
```rust
use async_trait::async_trait;

struct Foo;

#[async_trait]
trait T1 {
    type T;

    async fn f1(&self, t: Self::T) -> Self::T;
}

#[async_trait]
trait T2: T1 {
    async fn f2(&self, t: Self::T) -> Self::T;
}

#[async_trait]
impl T1 for Foo {
    type T = Foo;

    async fn f1(&self, t: Self::T) -> Self::T {
        t
    }
}

#[async_trait]
impl T2 for Foo {
    async fn f2(&self, t: Self::T) -> Self::T {
        self.f1(t).await
    }
}
```
rust compiler generates these errors:
```
error[E0576]: cannot find associated type `T` in trait `T2`
  --> experiments/src/main.rs:32:33
   |
32 |     async fn f2(&self, t: Self::T) -> Self::T {
   |                                 ^ not found in `T2`

error[E0576]: cannot find associated type `T` in trait `T2`
  --> experiments/src/main.rs:32:45
   |
32 |     async fn f2(&self, t: Self::T) -> Self::T {
   |                                             ^ not found in `T2`

error: aborting due to 2 previous errors
```
`cargo expand` output shows that attribute replaces associated type to its code:
```rust
async fn __f2(_self: &Foo, t: <Foo as T2>::T) -> <Foo as T2>::T {
    _self.f1(t).await
}
```
If I replace `Self::T` with `<Self as T1>::T` rustc will be happy.

**Comment by federico-terzi** [comment]
I'm currently facing the same issue

**Comment by dtolnay** [comment]
Fixed by async-trait 0.1.43.
